src/main/java/com/segment/analytics/android/integrations/firebase/FirebaseIntegration.java

```diff
@@ -3,88 +3,242 @@
+          activityLabel);
+    } catch (PackageManager.NameNotFoundException e) {
+      throw new AssertionError("Activity Not Found: " + e.toString());
     }
-  };
-
-  private static final String FIREBASE_ANALYTICS_KEY = "Firebase Analytics";
-
-  public FirebaseIntegration(Context context) {
-    mFirebaseAnalytics = FirebaseAnalytics.getInstance(context);
   }
 
-  @Override public void onActivityStarted(Activity activity) {
-    super.onActivityStarted(activity);
+  @Override
+  public void identify(IdentifyPayload identify) {
+    super.identify(identify);
 
+    if (!isNullOrEmpty(identify.userId())) {
+      firebaseAnalytics.setUserId(identify.userId());
+    }
+    Map<String, Object> traits = identify.traits();
+    for (Map.Entry<String, Object> entry : traits.entrySet()) {
+      String trait = entry.getKey();
+      trait = trait.trim().replaceAll(" ", "_");
+      if (trait.length() > 40) {
+        trait = trimKey(trait);
+      }
+      String formattedValue;
+      if (entry.getValue() instanceof Date) {
+        Date value = (Date) entry.getValue();
+        formattedValue = formatDate(value);
+      } else {
+        formattedValue = String.valueOf(entry.getValue());
+      }
+      firebaseAnalytics.setUserProperty(trait, formattedValue);
+      logger.verbose("firebaseAnalytics.setUserProperty(%s, %s);", trait, formattedValue);
+    }
   }
 
-  @Override public void onActivityStopped(Activity activity) {
-    super.onActivityStopped(activity);
+  @Override
+  public void track(TrackPayload track) {
+    super.track(track);
 
+    String event = track.event();
+    String eventName = mapEvent(event);
+    Properties properties = track.properties();
+    Bundle formattedProperties = formatProperties(properties);
+    firebaseAnalytics.logEvent(eventName, formattedProperties);
+    logger.verbose("firebaseAnalytics.logEvent(%s, %s);", eventName, formattedProperties);
   }
 
-  @Override public void identify(IdentifyPayload identify) {
-    super.identify(identify);
-
-    if (!isNullOrEmpty(identify.userId())) {
-      mFirebaseAnalytics.setUserId(identify.userId());
+  private String mapEvent(String event) {
+    String eventName = event;
+    if (eventMapper.containsKey(eventName)) {
+      eventName = eventMapper.get(eventName);
     }
+    eventName = eventName.trim().replaceAll(" ", "_");
+    if (eventName.length() > 40) {
+      eventName = trimKey(eventName);
+    }
+    return eventName;
+  }
 
-    // mFirebaseAnalytics.setUserProperty(key, value);
-
+  private Bundle formatProperties(Properties properties) {
+    Bundle bundle = new Bundle();
+    if ((properties.revenue() != 0 || properties.total() != 0)
+        && isNullOrEmpty(properties.currency())) {
+      bundle.putString(Param.CURRENCY, "USD");
+    }
+    for (Map.Entry<String, Object> entry : properties.entrySet()) {
+      String property = entry.getKey();
+      property = mapProperty(property);
+      if (entry.getValue() instanceof Integer) {
```

Comment:
Instead of calling `entry.getValue()` everywhere, this should be done once and stored as a variable as well.

e.g. with the first block
```
Object value = entry.getValue();
if (value instanceof Integer) {
  int intValue = (int) entry.getValue();
  bundle.putInt(property, intValue);
  ...
}
...
```